### PR TITLE
doc: producer event push sample

### PR DESCRIPTION
**Pull request by johanandren**

### CODE REVIEW

File: samples/grpc/local-drone-control-scala/src/main/scala/drones/Drone.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        }
+      }
+  }
+
+  private def handleEvent(state: State, event: Event): State = event match {
+    case PositionUpdated(newPosition) =>
+      val newHistoricalPositions = state.currentPosition match {
+        case Some(position) =>
+          val withPreviousPosition = state.historicalPositions.enqueue(position)
+          if (withPreviousPosition.size > LocationHistoryLimit)
```
Comment (patriknw) [inline samples/grpc/local-drone-control-scala/src/main/scala/drones/Drone.scala:89]: Not important, but I think we have "debated" this before 😉 . Queue isn't optimal for `size`. I prefer Vector for such thing. `tail` of Vector isn't too bad.

File: samples/grpc/local-drone-control-scala/src/main/scala/drones/Drone.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+          // same grid location as before
+          Effect
+            .persist(PositionUpdated(position))
+            .thenReply(replyTo)(_ => Done)
+        } else {
+          // new grid location
+          Effect
+            .persist(
+              PositionUpdated(position),
+              CoarseGrainedLocationChanged(newCoarseGrainedLocation))
```
Comment (patriknw) [inline samples/grpc/local-drone-control-scala/src/main/scala/drones/Drone.scala:78]: later, to illustrate starting from snapshots we could snapshot when there is a CoarseGrainedLocationChanged

**Review by patriknw — COMMENTED** [review]
looking good

File: samples/grpc/local-drone-control-scala/src/main/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  actor.provider = cluster
+  remote.artery.canonical.hostname = "127.0.0.1"
+
+  http.server.enable-http2 = on
+
+  persistence.r2dbc {
+    query {
+      refresh-interval = 500 millis
+      # reducing this to have quicker test, triggers backtracking earlier
+      backtracking.behind-current-time = 3 seconds
```
[the remark this answers is not in the document]
Comment (johanandren) [reply samples/grpc/local-drone-control-scala/src/main/resources/application.conf:14]: Yeah, copy pasta, I'll go through and make sure to drop

File: samples/grpc/local-drone-control-scala/src/main/resources/application.conf [hunk 4]
```diff
@@ -0,0 +1,44 @@
+include "h2-default-projection-schema.conf"
+# FIXME break this up in different files like the other samples?
+akka {
+  log-level = DEBUG
```
Comment (patriknw) [inline samples/grpc/local-drone-control-scala/src/main/resources/application.conf:4]: typo: loglevel

Comment (johanandren) [reply samples/grpc/local-drone-control-scala/src/main/resources/application.conf:4]: One day I will learn which it is...

File: samples/grpc/restaurant-drone-deliveries-service-scala/src/main/scala/central/deliveries/Main.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    AkkaManagement(system).start()
+    ClusterBootstrap(system).start()
+
+    val interface = system.settings.config.getString("restaurant-drone-deliveries-service.grpc.interface")
+    val port = system.settings.config.getInt("restaurant-drone-deliveries-service.grpc.port")
+
+    // consumer runs gRPC server accepting pushed events from producers
+    val streamId = "drone-events"
+    // FIXME we need to be able to pass protobuf descriptors for the gRPC journal via destination or as param to grpcServiceHandler here
+    //       right now failing on deserialization, also probably missing some docs on wire format for producer push
```
Comment (johanandren) [inline samples/grpc/restaurant-drone-deliveries-service-scala/src/main/scala/central/deliveries/Main.scala:43]: Fix with protobuf descriptor support added in https://github.com/akka/akka-projection/pull/957

**Comment by johanandren** [comment]
This now covers producer push. I think we can add events in the other direction as a separate PR on top.

File: samples/grpc/local-drone-control-scala/src/main/resources/persistence.conf [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  persistence {
+    journal {
+      plugin = "akka.persistence.r2dbc.journal"
+    }
+    snapshot-store {
+      plugin = "akka.persistence.r2dbc.snapshot"
+    }
+
+    r2dbc {
+      # FIXME mostly here to keep from lots of debug logs on start
```
Comment (patriknw) [inline samples/grpc/local-drone-control-scala/src/main/resources/persistence.conf:11]: can that be filtered with logback instead, or should something be changed with the logging?

File: samples/grpc/local-drone-control-scala/src/main/scala/local/drones/Drone.scala [hunk 7]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
```
Comment (patriknw) [inline samples/grpc/local-drone-control-scala/src/main/scala/local/drones/Drone.scala:2]: we can remove Lightbend copyright in the samples

**Review by patriknw — COMMENTED** [review]
looking good, just some minor things and then this can be merged before adding more features

File: samples/grpc/local-drone-control-scala/src/main/resources/persistence.conf (hunk 6 above)
Comment (johanandren) [reply samples/grpc/local-drone-control-scala/src/main/resources/persistence.conf:11]: On second thought I think we might not want all those topics for the small-local-drone-control service anyway, so maybe a low number here would be good. Just limiting resources.

File: samples/grpc/local-drone-control-scala/src/main/scala/local/drones/DroneEvents.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        settings = None,
+        sourceProvider = EventSourcedProvider.eventsBySlices[Drone.Event](
+          system,
+          R2dbcReadJournal.Identifier,
+          eventProducer.eventProducerSource.entityType,
+          0,
+          1023),
+        handler = eventProducer.handler()))
+
+    system.systemActorOf(behavior, "DroneEventPusher")
```
[the remark this answers is not in the document]
Comment (johanandren) [reply samples/grpc/local-drone-control-scala/src/main/scala/local/drones/DroneEvents.scala:70]: 👍 or require `SpawnProtocol` as system root

[the remark this answers is not in the document]
Comment (johanandren) [reply samples/grpc/local-drone-control-scala/src/main/scala/local/drones/DroneEvents.scala]: Changed to return Behaviour and Main using SpawnProtocol as root to start it (but maybe root actor bootstrap would align better with our Akka style guide, not sure, let's revisit separately if you think so)

File: samples/grpc/local-drone-control-scala/src/main/resources/persistence.conf (hunk 6 above)
Comment (johanandren) [reply samples/grpc/local-drone-control-scala/src/main/resources/persistence.conf]: Just the one projection running, so changed it to 1

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**